mlir/lib/ExecutionEngine/ROCm/BackendUtils.cpp

```diff
@@ -39,19 +39,106 @@
 // TODO: remove this once the rocm_agent_enumerator is ready
 #include "hip/hip_runtime.h"
 
+#include <cstring>
 #include <mutex>
 #include <numeric>
 
 using namespace mlir;
 
 static constexpr const char kRunnerProgram[] = "mlir-rocm-runner";
 static constexpr const char kRocmAgentEnumerator[] = "rocm_agent_enumerator";
+static constexpr const char gcnArchDelimiter[] = ":";
+static constexpr const char targetTriple[] = "amdgcn-amd-amdhsa";
+
+namespace {
+// TODO: remove this once the rocm_agent_enumerator is ready
+void getGpuGCNArchName(hipDevice_t device, std::string &gcnArchName) {
+  hipDeviceProp_t props;
+  hipError_t result = hipGetDeviceProperties(&props, device);
+  if (result != hipSuccess) {
+    gcnArchName = "";
+    return;
+  }
+
+  const char *pArchName = props.gcnArchName;
+  gcnArchName.assign(pArchName);
+}
+
+// This function convert the arch name to target features:
+// gfx908:sramecc+:xnack- to +sramecc,-xnack
+std::string getFeaturesFromArchName(const std::string &gcnArchName) {
+  std::string features;
+
+  // First step: get rid of the arch name in feature string
+  std::size_t firstSeperatorLoc = gcnArchName.find(gcnArchDelimiter);
+  if (firstSeperatorLoc == std::string::npos) {
+    return features;
+  }
+  std::string gcnArchFeature = gcnArchName.substr(firstSeperatorLoc);
+
+  // Second step: put each feature name to the vector
+  std::string token;
+  std::vector<std::string> featureTokens;
+  auto convertFeatureToken = [](std::string &token) {
+    if (token.back() != '+' && token.back() != '-') {
+      llvm::errs() << "malformed token: " << token << ", must end with +/-.\n";
+      return token;
```

Comment:
Is this the right thing to do with malformed tokens?

Reply:
Same question as above, do you see a better way to handle those? I have limited knowledge on a better to do error handling in general and typically use a `llvm::errs()`

Reply:
Having looked into the context, we probably want to use the idioms in the [failable constructor section](https://llvm.org/docs/ProgrammersManual.html#fallible-constructors) to report bad  values.

Reply:
Thanks for the pointer. Looking at the description and then the context, I don't believe either case we discuss about is recoverable errors. If this feature string is malformed or if the hip API is not able to successfully to return a good archname then the system must be very messed up. I'm pretty sure in both scenarios it soon evolve into a crash. So I'd just create `llvm_unreachable` statement on those.

Reply:
Ah, makes sense. Yeah, so let's just `unreachable()` this and hope that people who're taking, say, command-line input validate it before it gets to us

Reply:
(That is, if there's any environment where the feature string is being typed in by a user at the command line, there should be a nice error message before we get to this point - if there isn't one, let's add it?)

Reply:
So, the only possibility that a user could have a chance to input this feature string is by using the option in the `mlir-rocm-runner` (which I'm not sure if anyone ever exercised it). Now the embarrassing fact is that in order to know that the string is mal-formed, it has to be processed in the same way as we process the archname in here. Process this thing twice doesn't seem to be a good idea to me as well.

What I may end up doing is to add more context in the unreachable error string + the option description string itself, just so that the user knows giving this option a random string won't be able to work.

Reply:
I think what this calls for is better error reporting. This code is, in effect, part of our option validation pass, and option validation should be structured to report failure in a meaningful way, which, I'd argue, means propagating these sorts of errors up to the entry point.

Reply:
Okay, I was thinking in perspective of the runner. Now thinking from the static library, we sure don't want to be the only solver crash in applicability call. So that make sense. it make better sense to move this processing function in `Conv2dGenerator` (instead of constructor here) then.

Reply:
Addressed in 42f7e7c. Now IsaNameParser (instead of the constructor) is the component responsible for the parsing and it is status chained.